File: binance4j-savings/src/test/java/com/binance4j/savings/client/SavingsClientTest.java
package com.binance4j.savings.client;

import java.util.List;

import org.junit.jupiter.api.Test;

import com.binance4j.core.exception.ApiException;
import com.binance4j.core.test.CustomTest;
import com.binance4j.savings.dto.FixedProjectType;
import com.binance4j.savings.dto.LendingType;
import com.binance4j.savings.dto.ProductType;
import com.binance4j.savings.param.ChangePositionParams;
import com.binance4j.savings.param.FixedProjectListParams;
import com.binance4j.savings.param.FixedProjectPositionParams;
import com.binance4j.savings.param.FixedPurchaseParams;
import com.binance4j.savings.param.FlexiblePurchaseParams;
import com.binance4j.savings.param.LendingParams;
import com.binance4j.savings.param.PurchaseQuotaParams;
import com.binance4j.savings.param.RedemptionParams;
import com.binance4j.savings.param.RedemptionQuotaParams;
import com.fasterxml.jackson.databind.DeserializationFeature;

public class SavingsClientTest extends CustomTest {
	SavingsClient client = new SavingsClient(key, secret);

	public SavingsClientTest() {
		client.getMapper().configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, true);
		client.getMapper().configure(DeserializationFeature.FAIL_ON_MISSING_CREATOR_PROPERTIES, true);
	}

	@Test
	void testGetAccount() throws ApiException {
		testNotThrow(client.getAccount());
	}

	@Test
	void testGetFixedProjects() throws ApiException {
		testNotThrow(client.getFixedProjects(new FixedProjectListParams(FixedProjectType.CUSTOMIZED_FIXED)));
		testNotThrow(client.getFixedProjects(new FixedProjectListParams(FixedProjectType.ACTIVITY)));
	}

	@Test
	void testGetFlexibleProducts() throws ApiException {
		client.getMapper().configure(DeserializationFeature.FAIL_ON_MISSING_CREATOR_PROPERTIES, false);
		testHasNulls(client.getFlexibleProducts(), List.of("tierAnnualInterestRate"), true);
		client.getMapper().configure(DeserializationFeature.FAIL_ON_MISSING_CREATOR_PROPERTIES, true);
	}

	@Test
	void testGetInterests() throws ApiException {
		testNotThrow(client.getInterests(new LendingParams(LendingType.DAILY)));
		testNotThrow(client.getInterests(new LendingParams(LendingType.ACTIVITY)));
		testNotThrow(client.getInterests(new LendingParams(LendingType.CUSTOMIZED_FIXED)));
	}

	@Test
	void testGetPurchases() throws ApiException {
		client.getMapper().configure(DeserializationFeature.FAIL_ON_MISSING_CREATOR_PROPERTIES, false);
		testNotThrow(client.getPurchases(new LendingParams(LendingType.DAILY)));
		testNotThrow(client.getPurchases(new LendingParams(LendingType.ACTIVITY)));
		testNotThrow(client.getPurchases(new LendingParams(LendingType.CUSTOMIZED_FIXED)));
		client.getMapper().configure(DeserializationFeature.FAIL_ON_MISSING_CREATOR_PROPERTIES, true);
	}

	@Test
	void testGetRedemptions() throws ApiException {
		testNotThrow(client.getRedemptions(new LendingParams(LendingType.DAILY)));
		testNotThrow(client.getRedemptions(new LendingParams(LendingType.ACTIVITY)));
		testNotThrow(client.getRedemptions(new LendingParams(LendingType.CUSTOMIZED_FIXED)));
	}

	@Test
	void testGetLeftDailyFlexiblePurchaseQuota() throws ApiException {
		client.getMapper().configure(DeserializationFeature.FAIL_ON_MISSING_CREATOR_PROPERTIES, false);
		var productId = client.getFlexibleProducts().sync().get(0).productId();
		testNotThrow(client.getLeftDailyFlexiblePurchaseQuota(new PurchaseQuotaParams(productId)));
		client.getMapper().configure(DeserializationFeature.FAIL_ON_MISSING_CREATOR_PROPERTIES, true);
	}

	@Test
	void testGetLeftDailyRedemptionQuota() throws ApiException {
		client.getMapper().configure(DeserializationFeature.FAIL_ON_MISSING_CREATOR_PROPERTIES, false);
		var productId = client.getFlexibleProducts().sync().get(0).productId();
		testNotThrow(client.getLeftDailyRedemptionQuota(new RedemptionQuotaParams(productId, ProductType.FAST)));
		client.getMapper().configure(DeserializationFeature.FAIL_ON_MISSING_CREATOR_PROPERTIES, true);
	}

	// NOT TESTED

	// TODO @Test
	void testFixedToDailyPosition() throws ApiException {
		testNotThrow(client.fixedToDailyPosition(new ChangePositionParams("projectId", 0L)));
	}

	// TODO @Test
	void testGetFixedProjectPosition() throws ApiException {
		testNotThrow(client.getFixedProjectPosition(new FixedProjectPositionParams(null, null, null)));
	}

	// TODO @Test
	void testGetFlexibleProductPosition() throws ApiException {
		testNotThrow(client.getFlexibleProductPosition());
	}

	// TODO @Test
	void testPurchaseFixed() throws ApiException {
		testNotThrow(client.purchaseFixed(new FixedPurchaseParams("productId", 1L)));
	}

	// TODO @Test
	void testPurchaseFlexible() throws ApiException {
		testNotThrow(client.purchaseFlexible(new FlexiblePurchaseParams("productId", "amount")));
	}

	// TODO @Test
	void testRedeemFlexible() throws ApiException {
		testNotThrow(client.redeemFlexible(new RedemptionParams("productId", "amount", ProductType.FAST)));
	}
}

File: src/test/com/binance4j/connectors/savings/client/SavingsClientTest.kt
package com.binance4j.connectors.savings.client

import com.binance4j.connectors.connectors.Connectors
import com.binance4j.connectors.core.test.CustomTest
import com.binance4j.connectors.savings.dto.*
import org.junit.jupiter.api.Test

internal class SavingsClientTest : CustomTest() {
    private var client = Connectors.rest(key, secret).savings

    @Test
    fun `Get flexible products`() = assertDoesNotThrow(client.getFlexibleProducts())

    @Test
    fun `Get featured flexible products`() = assertDoesNotThrow(client.getFlexibleProducts(featured = Featured.TRUE))

    @Test
    fun `Get featured flexible products in an interval`() = assertDoesNotThrow(client.getFlexibleProducts(featured = Featured.TRUE, current = 1, size = limit))

    @Test
    fun `Get Left Daily Flexible Purchase Quota`() = assertDoesNotThrow(client.getLeftDailyFlexiblePurchaseQuota("ONG001"))

    //@Test TODO unquote to test
    //fun purchaseFlexible() = assertDoesNotThrow(client.purchaseFlexible("ONG001","1"))

    @Test
    fun `Get left daily redemption quota`() = assertDoesNotThrow(client.getLeftDailyRedemptionQuota("ONG001", ProductType.NORMAL))

    //@Test TODO unquote to test
    //fun redeemFlexible() = assertDoesNotThrow(client.redeemFlexible("ONG001", "1", ProductType.NORMAL))

    @Test
    fun `Get All Flexible Products Position`() = assertDoesNotThrow(client.getFlexibleProductPosition())

    @Test
    fun `Get Flexible Product Position`() = assertDoesNotThrow(client.getFlexibleProductPosition(asset))


    @Test
    fun `Get fixed projects`() = assertDoesNotThrow(
        client.getFixedProjects(
            type = FixedProjectType.CUSTOMIZED_FIXED,
            status = FixedProjectStatus.ALL,
            sortBy = FixedProjectSorting.DURATION
        )
    )

    //@Test TODO unquote to test
    //fun purchaseFixed() = assertDoesNotThrow(client.purchaseFixed("OP", 1L))

    @Test
    fun `Get Fixed Project Position`() = assertDoesNotThrow(client.getFixedProjectPosition(status = FixedProjectPositionStatus.HOLDING))

    @Test
    fun `Get account`() = assertDoesNotThrow(client.getAccount())

    @Test
    fun `Get all purchases`() = assertDoesNotThrow(client.getPurchases(LendingType.DAILY))

    @Test
    fun `Get all redemptions`() = assertDoesNotThrow(client.getRedemptions(LendingType.DAILY))

    @Test
    fun `Get all interests`() = assertDoesNotThrow(client.getInterests(LendingType.DAILY))

    //@Test TODO unquote to test
    //fun fixedToDailyPosition() = assertDoesNotThrow(client.fixedToDailyPosition("OP", 1L))
}